File: jruby/code/ruby_101/collect.rb
#---
# Excerpted from "Using JRuby",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material, 
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose. 
# Visit http://www.pragmaticprogrammer.com/titles/jruby for more book information.
#---
primes     = [1, 3, 5, 'seven']
composites = [2, 4, 6, 9, 10]
misc       = [0]

# Oops.  We should store 7 as a prime.  Let's fix that:
primes[3] = 7

numbers = {:primes => primes, :composites => misc}

# Oops.  We gave the wrong list for composites.  Let's fix that:
numbers[:composites] = composites

File: elixir16/code/protocols/collect.exs
#---
# Excerpted from "Programming Elixir ≥ 1.6",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material,
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose.
# Visit http://www.pragmaticprogrammer.com/titles/elixir16 for more book information.
#---
defimpl Collectable, for: Integer do
  def into(original) do
    {[], fn
      list, {:cont, x} -> [x|list]
      list, :done -> original ++ :lists.reverse(list)
      _, :halt -> :ok
    end}
  end
end

IO.inspect Enum.into(3, [])
